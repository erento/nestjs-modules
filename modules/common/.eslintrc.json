{
  "root": true,
  "ignorePatterns": [
    "projects/**/*"
  ],
  "overrides": [
    {
      "files": [
        "*.ts"
      ],
      "parserOptions": {
        "project": [
          "tsconfig.json"
        ],
        "createDefaultProgram": true
      },
      "extends": [
        "plugin:@erento/erento-rules/recommended"
      ],
      "plugins": [
        "@erento/erento-rules"
      ]
    }
  ]
}
