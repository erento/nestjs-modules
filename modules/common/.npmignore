src/
tsconfig.json
tslint.json
*.snap
*.test.d.ts
*.test.ts
*.test.js
