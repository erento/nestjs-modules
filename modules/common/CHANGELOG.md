<a name="7.5.0"></a>
# 7.5.0 (2023-10-23)
- added additional erento enum

<a name="7.4.0"></a>
# 7.4.0 (2023-09-07)
- added validation exception factory

<a name="7.3.0"></a>
# 7.3.0 (2023-03-07)
- added slugify text to utils.

<a name="7.2.0"></a>
# 7.2.0 (2023-02-09)
- added abstract pubsub pull subscriber

<a name="7.1.0"></a>
# 7.1.0 (2022-12-16)
- use trace&profiler only on beta/production, dev & test should be avoided

<a name="7.0.2"></a>
# 7.0.2 (2022-12-16)
- Another fix to avoid duplicate request id in the url in query param

<a name="7.0.1"></a>
# 7.0.1 (2022-12-16)
- Fix duplicate request id in the url in query param `_requestUniqueId`

<a name="7.0.0"></a>
# 7.0.0 (2022-12-15)
- Upgraded dependencies
- Do not run profiler & trace API during development (on localhost)

**Breaking Changes:**
- Introducing back `BugsnagIgnoreExceptionsInterceptor`
- Decorator `BugsnagIgnoreExceptions` is now again taking care of all exceptions

### New behavior:
**To silence exceptions within functions body use:**
```typescript
@BugsnagIgnoreExceptions([NotFoundException])
```

**To silence exceptions within authorization guard use:**
```typescript
@Auth(AuthorizationType.seller, {silent: true})
```

**To silence `UnauthorizedException` outside of authorization guard use also `BugsnagIgnoreExceptions`:**
```typescript
    @Auth(AuthorizationType.seller, {silent: true})
    @BugsnagIgnoreExceptions([UnauthorizedException])
    public getUserCreditsAsSeller (@Headers(TOKEN_USER_ID_HEADER) userId?: string): Promise<Credit> {
        if (!userId) {
            throw new UnauthorizedException('Unauthorized');
        }
```
In this case `Auth` decorator takes exceptions silent within authorization guard and `BugsnagIgnoreExceptions` within the function's body.

<a name="6.0.0"></a>
# 6.0.0 (2022-12-07)
**Breaking Changes:**
- removed `interceptAxiosRequestWithRequestId()` method and replaced by `RequestUniqueIdInterceptor`

Usage:
- import it in AppModule in provider and it will be used for all default HttpServices. In case the app is using `HttpModule.register`, you need to provide this interceptor again due to [this axios issue](https://github.com/axios/axios/issues/4938)

**Additional changes:**
- Added default constants as `COMMON_HEADERS`, `USER_AGENT`, etc... repeated in most of the service
- Added `onApplicationInit()` method to remove the same app boilerplate, should be called in `main.ts`
- Added `BasicAppService` and `BaseAppService` which is can be extended in case you need to extend `closeDatabaseConnection()`
- The unique ID is changed from `uuidv4()` to `${Environments.getApplicationName()}_${Environments.getVersion()}_${uuidv4()}`


<a name="5.5.0"></a>
# 5.5.0 (2022-09-28)
- fixed BugsnagIgnoreExceptions behaviour
- removed BugsnagInterceptor

<a name="5.4.0"></a>
# 5.4.0 (2022-09-28)
Added validation pipes:
- boolean-validation.pipe.ts
- iso-date-validation.pipe.ts
- optional-boolean-validation.pipe.ts
- optional-int-validation.pipe.ts
- radius-validation.pipe.ts
- string-list-validation.pipe.ts

<a name="5.3.0"></a>
# 5.3.0 (2022-08-22)
- added sorting param pipe

<a name="5.2.0"></a>
# 5.2.0 (2022-03-25)
- change `AuthorizationGuard` to return `UnauthorizedException`

<a name="5.1.1"></a>
# 5.1.1 (2022-03-25)
- fix wrong usage of rxjs `of()` method

<a name="5.1.0"></a>
# 5.1.0 (2022-03-07)
- added universal request unique id interceptor

<a name="5.0.0"></a>
# 5.0.0 (2022-03-7)
- updated dependencies to latest possible versions
- raised required node >=16.0.0 & npm to >=8.0.0
- replaced TSLint with ESLint
- fixed new linting issues
- updated husky & pre-commit hooks
- changed constructor arguments of LocaleParamPipe
