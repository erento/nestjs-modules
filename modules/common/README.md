# NestJS common module

Common and utility functions used across our NestJS projects.

## Stay in touch

* [Erento's developers](mailto:developers@erento.com) 

## License

This module is [MIT licensed](LICENSE.md).
