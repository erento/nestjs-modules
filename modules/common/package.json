{
  "name": "@erento/nestjs-common",
  "version": "1.0.0",
  "description": "NestJS module with some shared utilities",
  "main": "dist/public_api.js",
  "types": "dist/public_api.d.ts",
  "scripts": {
    "build": "tsc -p tsconfig.json",
    "lint": "tslint -c tslint.json --fix -e \"**/node_modules/**\" -p ./tsconfig.json",
    "lint-staged": "lint-staged",
    "prebuild": "rm -rf dist && npm run lint && npm run test",
    "test": "jest"
  },
  "keywords": [],
  "author": "Erento GmbH <developers@erento.com> (https://www.erento.com)",
  "repository": {
    "type": "git",
    "url": "https://github.com/erento/nestjs-modules.git"
  },
  "license": "MIT",
  "peerDependencies": {
    "@bugsnag/js": "^6.5.2",
    "@bugsnag/plugin-express": "^6.5.1",
    "@nestjs/common": "^6.11.11",
    "@nestjs/core": "^6.11.11",
    "express-http-context": "^1.2.3",
    "json-stringify-safe": "^5.0.1",
    "typescript": ">=3.5"
  },
  "devDependencies": {
    "@types/jest": "26.0.3",
    "@types/node": "14.0.14",
    "husky": "4.2.5",
    "jest": "26.1.0",
    "lint-staged": "10.2.11",
    "prettier": "2.0.5",
    "reflect-metadata": "0.1.13",
    "ts-jest": "25.5.1",
    "tslint": "6.1.2",
    "tslint-eslint-rules": "5.4.0",
    "typescript": "3.8.3"
  },
  "dependencies": {
    "@bugsnag/js": "6.5.2",
    "@bugsnag/plugin-express": "6.5.1",
    "@nestjs/common": "6.11.11",
    "@nestjs/core": "6.11.11",
    "express-http-context": "1.2.3",
    "json-stringify-safe": "5.0.1"
  },
  "precommit": [
    "lint-staged"
  ],
  "lint-staged": {
    "src/**/*.ts": [
      "npm run lint",
      "git add"
    ]
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ]
  }
}
