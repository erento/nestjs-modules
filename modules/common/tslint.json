{
    "rulesDirectory": [
        "node_modules/tslint-eslint-rules/dist/rules"
    ],
    "defaultSeverity": "error",
    "rules": {
        "adjacent-overload-signatures": true,
        "align": false,
        "array-bracket-spacing": [
            true,
            "never"
        ],
        "array-type": [
            true,
            "array"
        ],
        "arrow-parens": [
            true,
            "ban-single-arg-parens"
        ],
        "arrow-return-shorthand": false,
        "await-promise": false,
        "ban-comma-operator": true,
        "ban-ts-ignore": true,
        "ban-types": true,
        "binary-expression-operand-order": false,
        "block-spacing": [
            true,
            "always"
        ],
        "brace-style": [
            true,
            "1tbs"
        ],
        "class-name": true,
        "comment-format": false,
        "comment-type": false,
        "callable-types": true,
        "curly": true,
        "encoding": true,
        "eofline": true,
        "forin": true,
        "file-header": false,
        "file-name-casing": false,
        "function-constructor": true,
        "increment-decrement": false,
        "interface-over-type-literal": false,
        "import-blacklist": true,
        "import-spacing": true,
        "indent": [
            true,
            "spaces",
            4
        ],
        "jsdoc-format": [true, "check-multiline-start"],
        "label-position": true,
        "linebreak-style": [
            true,
            "LF"
        ],
        "match-default-export-name": true,
        "max-classes-per-file": [
            true,
            1
        ],
        "max-file-line-count": [
            true,
            600
        ],
        "max-line-length": [
            true,
            140
        ],
        "member-access": true,
        "member-ordering": [
            true,
            {
                "order": [
                    "public-static-field",
                    "public-static-method",
                    "protected-static-field",
                    "protected-static-method",
                    "private-static-field",
                    "private-static-method",
                    "public-instance-field",
                    "protected-instance-field",
                    "private-instance-field",
                    "public-constructor",
                    "protected-constructor",
                    "private-constructor",
                    "public-instance-method",
                    "protected-instance-method",
                    "private-instance-method"
                ]
            }
        ],
        "new-parens": true,
        "newline-before-return": false,
        "newline-per-chained-call": false,
        "no-angle-bracket-type-assertion": false,
        "no-arg": true,
        "no-bitwise": true,
        "no-boolean-literal-compare": false,
        "no-conditional-assignment": true,
        "no-consecutive-blank-lines": [
            true,
            1
        ],
        "no-console": [
            true,
            "debug",
            "time",
            "timeEnd",
            "trace"
        ],
        "no-constant-condition": true,
        "no-construct": true,
        "no-control-regex": true,
        "no-debugger": true,
        "no-default-export": true,
        "no-default-import": true,
        "no-duplicate-case": true,
        "no-duplicate-imports": true,
        "no-duplicate-variable": [true, "check-parameters"],
        "no-duplicate-super": true,
        "no-duplicate-switch-case": true,
        "no-dynamic-delete": false,
        "no-empty": false,
        "no-empty-character-class": true,
        "no-empty-interface": false,
        "no-eval": true,
        "no-ex-assign": true,
        "no-extra-boolean-cast": true,
        "no-extra-semi": true,
        "no-for-in-array": true,
        "no-floating-promises": false,
        "no-implicit-dependencies": [true, "dev"],
        "no-import-side-effect": false,
        "no-inferrable-types": false,
        "no-inferred-empty-object-type": true,
        "no-inner-declarations": [
            true,
            "both"
        ],
        "no-internal-module": true,
        "no-invalid-regexp": true,
        "no-invalid-template-strings": true,
        "no-invalid-this": false,
        "no-irregular-whitespace": true,
        "no-magic-numbers": false,
        "no-mergeable-namespace": true,
        "no-misused-new": true,
        "no-multi-spaces": true,
        "no-namespace": [
            true,
            "allow-declarations"
        ],
        "no-non-null-assertion": true,
        "no-null-keyword": false,
        "no-parameter-properties": false,
        "no-parameter-reassignment": true,
        "no-redundant-jsdoc": false,
        "no-reference": true,
        "no-reference-import": true,
        "no-regex-spaces": true,
        "no-require-imports": true,
        "no-return-await": true,
        "no-shadowed-variable": true,
        "no-sparse-arrays": true,
        "no-string-literal": false,
        "no-string-throw": true,
        "no-submodule-imports": false,
        "no-switch-case-fall-through": true,
        "no-this-assignment": false,
        "no-trailing-whitespace": true,
        "no-unbound-method": true,
        "no-unexpected-multiline": true,
        "no-unnecessary-callback-wrapper": true,
        "no-unnecessary-class": false,
        "no-unnecessary-initializer": true,
        "no-unnecessary-qualifier": false,
        "no-unnecessary-type-assertion": false,
        "no-unsafe-any": false,
        "no-unsafe-finally": true,
        "no-unused-expression": true,
        "no-var-keyword": true,
        "no-var-requires": true,
        "no-void-expression": [true, "ignore-arrow-function-shorthand"],
        "number-literal-format": true,
        "object-literal-key-quotes": [
            true,
            "consistent"
        ],
        "object-literal-shorthand": true,
        "object-literal-sort-keys": false,
        "one-line": [
            true,
            "check-catch",
            "check-else",
            "check-finally",
            "check-open-brace",
            "check-whitespace"
        ],
        "one-variable-per-declaration": true,
        "object-curly-spacing": [
            true,
            "never"
        ],
        "ordered-imports": [
            true,
            {
                "grouped-imports": false,
                "import-sources-order": "case-insensitive",
                "module-source-path": "full",
                "named-imports-order": "case-insensitive"
            }
        ],
        "prefer-conditional-expression": true,
        "prefer-const": true,
        "prefer-for-of": true,
        "prefer-function-over-method": false,
        "prefer-method-signature": true,
        "prefer-readonly": false,
        "prefer-object-spread": true,
        "prefer-switch": ["true", { "min-cases": 3 }],
        "prefer-template": [true, "allow-single-concat"],
        "promise-function-async": false,
        "quotemark": [
            true,
            "single",
            "jsx-double",
            "avoid-escape"
        ],
        "radix": true,
        "restrict-plus-operands": false,
        "return-undefined": true,
        "semicolon": [
            true,
            "always"
        ],
        "sort-imports": [false],
        "space-before-function-paren": true,
        "space-in-parens": [true, "never"],
        "space-within-parens": [true, 0],
        "switch-default": true,
        "switch-final-break": [true, "always"],
        "strict-boolean-expressions": false,
        "strict-type-predicates": true,
        "ter-arrow-body-style": [
            true,
            "as-needed",
            {
                "requireReturnForObjectLiteral": true
            }
        ],
        "ter-arrow-parens": [
            true,
            "as-needed"
        ],
        "ter-arrow-spacing": [
            true
        ],
        "ter-func-call-spacing": [true, "never"],
        "ter-indent": [
            true,
            4,
            {
                "FunctionDeclaration": {
                    "parameters": 1,
                    "body": 1
                },
                "FunctionExpression": {
                    "parameters": 1,
                    "body": 1
                },
                "SwitchCase": 1,
                "MemberExpression": 1,
                "CallExpression": {
                    "arguments": 1
                },
                "VariableDeclarator": {
                    "var": 1,
                    "let": 1,
                    "const": 1
                }
            }
        ],
        "ter-no-irregular-whitespace": true,
        "ter-no-sparse-arrays": true,
        "ter-no-tabs": true,
        "ter-prefer-arrow-callback": [
            true
        ],
        "trailing-comma": [
            true,
            {
                "multiline": {
                    "objects": "always",
                    "arrays": "always",
                    "functions": "always",
                    "typeLiterals": "always"
                },
                "singleline": {
                    "objects": "never",
                    "arrays": "never",
                    "functions": "never",
                    "typeLiterals": "never"
                },
                "esSpecCompliant": true
            }
        ],
        "triple-equals": [
            true,
            "allow-null-check"
        ],
        "type-literal-delimiter": true,
        "typedef": [
            true,
            "arrow-call-signature",
            "array-destructuring",
            "arrow-parameter",
            "call-signature",
            "member-variable-declaration",
            "object-destructuring",
            "parameter",
            "property-declaration",
            "variable-declaration"
        ],
        "typedef-whitespace": [
            true,
            {
                "call-signature": "nospace",
                "index-signature": "nospace",
                "parameter": "nospace",
                "property-declaration": "nospace",
                "variable-declaration": "nospace"
            },
            {
                "call-signature": "onespace",
                "index-signature": "onespace",
                "parameter": "onespace",
                "property-declaration": "onespace",
                "variable-declaration": "onespace"
            }
        ],
        "unified-signatures": true,
        "unnecessary-bind": true,
        "unnecessary-constructor": true,
        "use-default-type-parameter": true,
        "use-isnan": true,
        "valid-jsdoc": [true, {
            "prefer": {
                "return": "returns"
            },
            "requireParamDescription": false,
            "requireReturnDescription": false
        }],
        "valid-typeof": true,
        "variable-name": [true, "ban-keywords", "check-format", "allow-leading-underscore"],
        "whitespace": [
            true,
            "check-branch",
            "check-decl",
            "check-operator",
            "check-preblock",
            "check-rest-spread",
            "check-separator",
            "check-type",
            "check-type-operator",
            "check-typecast"
        ]
    }
}
