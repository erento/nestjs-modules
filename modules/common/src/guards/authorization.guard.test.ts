import {ExecutionContext} from '@nestjs/common';
import {AuthorizationGuard} from './authorization.guard';

const getExecutionContext: (req: any) => ExecutionContext = (req: any): ExecutionContext => <any> {
    switchToHttp: (): any => {
        return {getRequest: (): any => req};
    },
    handler: {},
};

describe('Authorization Guard', (): void => {
    it('Should authorize when token is not passed', (): void => {
        const reflector: any = jest.fn();
        reflector.mockReturnValueOnce(undefined);
        const req: Request = <any> {
            headers: {},
        };
        expect(new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req))).toBe(true);
    });

    it('Should authorize when token is passed and matches req header', (): void => {
        const req: Request = <any> {
            headers: {
                'x-token-role': 'service',
            },
        };
        const reflector: any = jest.fn();
        reflector.mockReturnValueOnce('service');
        expect(new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req))).toBe(true);
    });

    it('Should authorize when token is passed as an array and matches req header', (): void => {
        const req: Request = <any> {
            headers: {
                'x-token-role': 'service',
            },
        };
        const reflector: any = jest.fn();
        reflector.mockReturnValueOnce(['x', 'y', 'service', 'z']);
        expect(new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req))).toBe(true);
    });

    it('Should authorize when req header is passed but no requirements needed', (): void => {
        const req: Request = <any> {
            headers: {
                'x-token-role': 'service',
            },
        };
        const reflector: any = jest.fn().mockReturnValueOnce(undefined);
        expect(new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req))).toBe(true);
    });

    it('Should not authorize when token does not matches req header', (): void => {
        const req: Request = <any> {
            headers: {
                'x-token-role': 'service',
            },
        };
        const reflector: any = jest.fn();
        reflector.mockReturnValueOnce('not matching');
        expect((): boolean => new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req)))
            .toThrowErrorMatchingSnapshot();
    });

    it('Should not authorize when array of tokens does not matches req header', (): void => {
        const req: Request = <any> {
            headers: {
                'x-token-role': 'service',
            },
        };
        const reflector: any = jest.fn();
        reflector.mockReturnValueOnce(['not matching', 'not-matching-2']);
        expect(
            (): boolean => new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req)),
        ).toThrowErrorMatchingSnapshot();
    });

    it('Should not authorize when no req header is passed', (): void => {
        const req: Request = <any> {
            headers: {},
        };
        const reflector: any = jest.fn();
        reflector.mockReturnValueOnce('not matching');
        expect((): boolean => new AuthorizationGuard(<any> {get: reflector}).canActivate(getExecutionContext(req)))
            .toThrowErrorMatchingSnapshot();
    });
});
