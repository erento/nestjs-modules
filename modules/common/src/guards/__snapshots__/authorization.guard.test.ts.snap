// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization Guard Should not authorize when array of tokens does not matches req header 1`] = `
Object {
  "error": "Not Authorized",
  "reason": "Provided token is \\"service\\"",
  "statusCode": 401,
}
`;

exports[`Authorization Guard Should not authorize when no req header is passed 1`] = `
Object {
  "error": "Not Authorized",
  "reason": "Provided token is \\"undefined\\"",
  "statusCode": 401,
}
`;

exports[`Authorization Guard Should not authorize when token does not matches req header 1`] = `
Object {
  "error": "Not Authorized",
  "reason": "Provided token is \\"service\\"",
  "statusCode": 401,
}
`;
